tests/components/statsd/test_init.py

```diff
@@ -8,133 +7,128 @@
-            host="host", port=123, prefix="foo"
+async def test_statsd_setup_full(hass):
+    """Test setup with all data."""
+    config = {"statsd": {"host": "host", "port": 123, "rate": 1, "prefix": "foo"}}
+    hass.bus.listen = MagicMock()
+    with patch("statsd.StatsClient") as mock_init:
+        assert await async_setup_component(hass, statsd.DOMAIN, config)
+
+        assert mock_init.call_count == 1
+        assert mock_init.call_args == mock.call(host="host", port=123, prefix="foo")
+
+    assert hass.bus.listen.called
+    assert EVENT_STATE_CHANGED == hass.bus.listen.call_args_list[0][0][0]
+
+
+async def test_statsd_setup_defaults(hass):
+    """Test setup with defaults."""
+    config = {"statsd": {"host": "host"}}
+
+    config["statsd"][statsd.CONF_PORT] = statsd.DEFAULT_PORT
+    config["statsd"][statsd.CONF_PREFIX] = statsd.DEFAULT_PREFIX
+
+    hass.bus.listen = MagicMock()
+    with patch("statsd.StatsClient") as mock_init:
+        assert await async_setup_component(hass, statsd.DOMAIN, config)
+
+        assert mock_init.call_count == 1
+        assert mock_init.call_args == mock.call(host="host", port=8125, prefix="hass")
+    assert hass.bus.listen.called
+
+
+async def test_event_listener_defaults(hass, mock_client):
+    """Test event listener."""
+    config = {"statsd": {"host": "host", "value_mapping": {"custom": 3}}}
+
+    config["statsd"][statsd.CONF_RATE] = statsd.DEFAULT_RATE
+
+    hass.bus.listen = MagicMock()
+    await async_setup_component(hass, statsd.DOMAIN, config)
+    assert hass.bus.listen.called
+    handler_method = hass.bus.listen.call_args_list[0][0][1]
+
+    valid = {"1": 1, "1.0": 1.0, "custom": 3, STATE_ON: 1, STATE_OFF: 0}
+    for in_, out in valid.items():
+        state = MagicMock(state=in_, attributes={"attribute key": 3.2})
+        handler_method(MagicMock(data={"new_state": state}))
+        mock_client.gauge.assert_has_calls(
+            [mock.call(state.entity_id, out, statsd.DEFAULT_RATE)]
         )
 
-        assert self.hass.bus.listen.called
-        assert EVENT_STATE_CHANGED == self.hass.bus.listen.call_args_list[0][0][0]
+        mock_client.gauge.reset_mock()
 
-    @mock.patch("statsd.StatsClient")
-    def test_statsd_setup_defaults(self, mock_connection):
-        """Test setup with defaults."""
-        config = {"statsd": {"host": "host"}}
+        assert mock_client.incr.call_count == 1
+        assert mock_client.incr.call_args == mock.call(
+            state.entity_id, rate=statsd.DEFAULT_RATE
+        )
+        mock_client.incr.reset_mock()
 
-        config["statsd"][statsd.CONF_PORT] = statsd.DEFAULT_PORT
-        config["statsd"][statsd.CONF_PREFIX] = statsd.DEFAULT_PREFIX
+    for invalid in ("foo", "", object):
+        handler_method(
+            MagicMock(data={"new_state": ha.State("domain.test", invalid, {})})
+        )
+        assert not mock_client.gauge.called
+        assert mock_client.incr.called
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6285789)